### DISCUSSION TITLE
What do you use Trail Sense for?

### DISCUSSION THREAD

**Discussion by kylecorry31** [kylecorry31/Trail-Sense#353] [New Feature Ideas]
I built Trail Sense because I wanted an app that I could depend on when hiking (or in the event I ever found myself in a wilderness survival situation). So far, I've used it for several hiking trips - going so far as to not even need to use a map on an overnight backpacking trip (I still brought it with me though), and I use it every day for weather forecasts.

**Comment by jer194** [comment]
I have just discovered your app, and I am just experimenting with it, discovering all of its features.
It makes me speechless, I have to say: clean UI, a cool toolset (eg even metal detector), lots of useful things, and even opensource!
(I like sensors ...)
Congatulations, and thank you!

**Reply by kylecorry31** [reply]
Thank you so much!!

**Reply by ybrhue** [reply]
Can you use trail sense to navigate existing tracks by importing one gpx file like in osmand?

**Reply by kylecorry31** [reply]
@ybrhue as of right now you can't, but I have issue #535 to track that as a future feature.

**Comment by Lee-Carre** [comment]
Were I still in my youth, then I would've used it for expeditions (Cubs, Scouts) and Geocaching. Plus, I'm sure it would've been handy while exploring Australia (Victoria) with family.

However, today, I find it useful for map ([OSM](//www.openstreetmap.org/)) surveying.

Part of why I like it, though, is that it gives a tech-infused nostalgia of weekend adventures.

**Reply by kylecorry31** [reply]
Thank you!

I hadn't really thought much about OSM surveying as a use case - how do you typically use Trail Sense for this? I imagine you either create beacons, backtrack paths, or record info shown on the navigation tab (location, bearing).

And also, would there be any features you would like to see added to Trail Sense that would make it more useful for map surveying?

**Reply by Lee-Carre** [reply]
This became rather longer than expected.

> how do you typically use Trail Sense for [map surveying]? I imagine you either create beacons, backtrack paths, or record info shown on the navigation tab (location, bearing)

Some context: while that may well have been the case many years ago (surveying on paper (or otherwise making notes), to then input via micro-computer after returning to base), there are now quite a few editors for nano-computers to enable real-time (or, at least, on-the-go) editing & upload of the resulting changeset. I started (as a newbie) with [StreetComplete](/streetcomplete/StreetComplete), but lately use [Vespucci](/MarcusWolschon/osmeditor4android) most commonly.

Vespucci, in particular, is very feature-rich (it's a fully-capable editor, for Android, intended for surveying). Thus, data like coordinates is all handled in-editor. That said, I use a variety of apps (I subscribe to the [POSIX philosophy](http://www.catb.org/esr/writings/taoup/) of modularity; each component should have one job in order to do that one job well), and I'll cite the relevant ones (I collect more than just map data, while out) in case you wish to take inspiration from them.

For recording where I've been (generally (moderate-resolution sampling), rather than capturing a (high-resolution) GNSS trace for upload) I use [TinyTravelTracker](/redfish64/TinyTravelTracker). Though, I might use a different package (but no idea which one, yet) when I have the infrastructure to run a backend logger, so that only a position-reporting app needs to run on the roving device itself.

I also haven't tinkered with Trail Sense as much as I should (I keep forgetting to experiment, while out), but a few features caught my attention so far:
* Inclinometer
* Bubble Level

The inclinometer would be useful for the following:
* measuring (numerically) the [incline](//wiki.openstreetmap.org/wiki/Key:incline) of (road)ways (since routing algorithms could make use of it) and the likes of (wheelchair-)ramps
* estimating the dimensions (especially height) of large structures (e.g. buildings) (really, any distances beyond what's practical for a tape-measure, until I get a laser rangefinder)
* measuring actual angles; non-rectangular buildings, tilt (CCTV, lighting (and the compass for pan direction, too)), roofs (viz with an apex, since for a fixed width the angle changes the roof-height (think of 3D renderings))

Re the Bubble Level, see also [Bubble](/woheller69/Level) & [Clinometer](/BasicAirData/Clinometer) (there's a few other rudimentary ones which I seldom use, but can cite if you wish).

If my device had a barometer, then I would definitely be making use of the weather(-prediction) features (autonomous computing for the win). So much so that I'll likely ensure that the next device, to replace this one, has a barometer & (ambient) thermometer. A precise barometer (down to thousandths of a millibar) is also rather useful for (more) accurately (than GNSS) measuring altitude/elevation, and thus the height of (human-accessible) features (buildings), without needing line-of-sight.

For measuring the height of open-roof buildings (public carparks), I'm tempted to use the cliff-height tool 😋.

Although it's still experimental, the Tide-prediction tool(s), since I live on [an island with very large tidal movements](//www.gov.je/Weather/Pages/Tides.aspx).

Being alerted before Sol is due to set. Though since, in Trail Sense, this is a perma-toggle (rather than only for the current day (or, more specifically, while I'm out)) I still use [Suntimes](/forrestguice/SuntimesWidget).

I can see the value of the lightning-distance tool (since it's real-time, autonomous, and works anywhere), but since I seldom venture far from civilisation then [Blitzortung Lightning Monitor](/wuan/bo-android) (else, [Web-based map](//map.blitzortung.org/)) serves well.

The Ruler for measuring smaller distances (and avoiding the need to always carry a separate physical ruler or tape-measure), such as [step-height](//taginfo.openstreetmap.org/keys/wheelchair:step_height). See also [Ruler](/SecUSo/privacy-friendly-ruler), [Ruler](/congshengwu/Ruler), & [Tape Measure](/SecUSo/privacy-friendly-tape-measure).

Beacons would be useful in the following cases:
* marking where I parked for easy return 🙂 (imagine it dark, cold, possibly wet, and returning along a different path to the original one)
* marking something which I intend (in the same session) to come back to (maybe it's non-trivial to map, or I'll need to read documentation, or it needs a lot of work to do properly)
* for areas with poor satellite reception (e.g. tree cover or inside buildings) being able to triangulate a reasonably-accurate position of a feature, from known-good positions
* when mapping paths, marking where the forks are (arguably a duplicate of the 2nd item in this list) since I can only do one at a time

Paths might be useful, but (for temporary reference) a similar feature exists in Vespucci, and for GNSS traces I use [OSMtracker](/labexp/osmtracker-android). Aside from previously mentioned apps (viz TinyTravelTracker), for general (day-to-day) navigation(+routing) I use [OsmAnd](/osmandapp/OsmAnd).

I can well imagine using the Packing Lists for more involved (e.g. by car, rather than on-foot) outings.

I'll try to remember to revisit this discussion, once I have more experience with Trail Sense, and thus a (hopefully) better answer.

> would there be any features you would like to see added to Trail Sense that would make it more useful for map surveying?

Great question. I'll have to think about this one; none come to mind right now. Many features already implemented in Trail Sense are ones which I would never have thought of; it's quite a toolset you've developed 🙂.

One thought, though it may be beyond the scope of Trail Sense (especially since internal GNSS receivers will likely be capable of higher-accuracy (dual-frequency, carrier phase, raw measurement output for SDR processing, RTK) in future), would be the option to have it use an external GNSS receiver for position. Simply being able to receive NMEA messages via a TCP port should be sufficient, rather than replicating what other software already does.

Not related to map-surveying, but since there is a moderate focus on astronomy within Trail Sense, then a display mode which avoids interfering with dark-adapted eyes. Red doesn't cause adapted eyes to lose their adaptation (hence red filters for lighting), black is obviously fine (as likely are dark shades of colours toward the red end of the spectrum). Definitely avoid white, blue, green. Dimmer is preferable to brighter, though text still has to be readable (but, that's part of the idea of using red; doesn't have to be super-dim then), but maybe increasing the font-size for that theme would suffice (preferably user-configurable; as a percentage increase from the system's configured default, rather than absolute overrides).

An even wilder idea (returning to surveying) would be to measure distance (in a single dimension) by computing linear displacement from the accelerometer data. A virtual, endless tape-measure of sorts, which should be more accurate over short distances than GNSS. Imagine lacking a laser rangefinder, and a sufficiently long tape-measure, while wanting to measure the length of a building-sized feature, or the width of a tunnel.
This might also help automate other tools (like the distance input for triangulation).
See [SmartNavi](/Phantast/smartnavi) for a proof-of-concept.

Along similar lines, being mindful of cases where integration of related tools (e.g. incliometer & bubble level) might enable usage of the whole-is-greater-than-sum-of-parts variety. Else simply automating some aspects (take inspiration from [POSIX usability design](http://www.catb.org/esr/writings/taouu/taouu.html) philosophy: never ask the user for input which the machine already has or can derive for itself).

A more specific (and less-challenging-to-implement) scenario might be to import (as beacons?) a set of places to visit, when doing more focused surveying especially over greater distances, as waypoints to survey during the session.
Context: often, I simply head out without much of a plan. In my area, there are very few local contributors, and thus a great deal still to be done. So, seldom need a plan, since there will be something easily findable which needs attention. However, on occasion, I do focus on specific types of targets (small amenities: post boxes, public telephones, defibrillators, recycling collection bins, vehicle charging stations). So, it would be helpful to able to avoid a memory-intensive general-navigation app (because I have enough running that memory-exhaustion is a risk, to the extent that I can't have a browser running lest it cause disruption (halting data-capture) in other apps) when all I really need is a geo-referenced check-list for the day. Don't need a map, don't need routing, just a direction & range marker; thus thinking that Trail Sense('s beacons) might be a suitable candidate). Besides resurveying amenities already in OSM which haven't been edited in a while, I also add missing features from what little info is made public (e.g. a list of the original set of public defibrillators); so, besides importing from well-known sources, being able to (manually-input if necessary) locations of not-yet-added features from a list, would be most helpful.
Obviously, with some ability to mark each as visited (or otherwise finished-with) in order to be directed to the next marker. Preferably automatically, based on my GNSS position being within a few metres of the marker, though also manually (for various reasons).

Other sources of inspiration:
* [PhyPhox](/phyphox/phyphox-android) (physics experiments toolkit)
* [Hikar](//gitlab.com/nickw1/Hikar) (augmented-reality; overlay OSM path data on the image from the rear camera)

**Reply by kylecorry31** [reply]
Thank you for all the details! I'll give all those apps a try and see if I can take inspiration from them. While not specifically catering to surveying, I'll keep that in mind while creating new features to see if there are small tweaks I could make to help out in this area (especially since I'm going to be moving the maps feature out of experimental soon - most of the PDF maps I use were created using OSM data)

I can definitely see areas where the inclinometer can be improved (which would also be useful for hiking/wilderness treks)

I do also have plans for an "astronomy" color theme (#210). 

Unfortunately, phone accelerometers are way too noisy to calculate even velocity, yet alone distance - but the pedometer could be used to get a rough estimate.

**Reply by Lee-Carre** [reply]
> …

All sounds good.

> I do also have plans for an "astronomy" color theme (#210).

👍 hadn't noticed that issue. Will follow.

> phone accelerometers are way too noisy to calculate even velocity, [l]et alone distance

Ah, wasn't aware of that problem. Nevermind, then. Laser-rangefinder it is.

> the pedometer could be used to get a rough estimate.

Having used some, before (in general, not for surveying), their accuracy also isn't sufficient (what exactly counts as a ‘step’, plus it's also dependent on consistent & accurate stride-length). Good enough for ‘how much have a walked, today’, but not for measurements requiring centimetre resolution.

Back to acquiring the proper tool of a laser measure 🙂.

> … the PDF maps I use were created using OSM …

Why PDF rather than an image? … I can only guess for vector rendering?

Now I'm curious why not use OsmAnd or Organic Maps, and preloaded with data for the relevant areas for them to use offline? Besides a task-specific UI (and identifying your own position), you get all the benefits of searching & routing.

**Reply by kylecorry31** [reply]
The PDF is used from CalTopo because it uses the Geospatial PDF metadata - which is used to auto-calibrate the map (the corners of the PDF image contain both pixel and geo coordinate information). I don't use vector/tile maps for the following reasons:
1. The maps in Trial Sense are based on a single map image rather than tiles/vector of a large area
2. There are already great apps for offline maps (ex. OsmAnd) which serve that purpose well (and I don't have the time/knowledge to reimplement them)
3. It is pretty easy to open a single image with Trail Sense (whether it's a JPG, PNG, PDF, etc), whereas loading map tiles is harder and requires a way to generate and load them onto the device (TS can't have internet permissions to gather tiles for itself)
4. So essentially, the maps in Trail Sense are designed around being able to take a picture of a physical map and use that - the PDFs just come as a bonus feature for higher accuracy

I do have #604 to add support, but I'm most likely not going to do that for the above reasons.

**Reply by Lee-Carre** [reply]
Interesting. Thanks for explaining. Wasn't aware of the metadata advantage.

I wasn't meaning to suggest tilesets (only single large images, like you say) or to directly use offline data from other apps. Reimplementing OsmAnd would be quite a mission.

My thinking-aloud question was re using those apps directly. I'm failing to see the gain of exporting from OSM to use that export in TS. Especially when using OA or OM yields good results.

Maybe my assumption is faulty: that the use-case for using an image as a custom map, is for specific sites whose own map is more detailed than OSM (and adding all that, even if licensing allows, to OSM would take longer).

Unless it's for purposes of integration; being able to use TS's many other features with an OSM-Carto background, instead of having to switch between multiple apps?

**Reply by kylecorry31** [reply]
It can be that the image is site specific and more detailed, but my main goal here was to be able to take a picture while on the trail and calibrate using that. If you already have an offline map downloaded, that is most likely the preferred solution - TS is a backup for when you only have access to a physical map or no internet.

Also, when I mentioned that I was using a PDF export of OSM, that was just one example of something I use - you can take a picture of a printed map or download any image or PDF for use in TS - which could be really useful for the average user who might just find a PDF on the trail's website or take a picture of a trail sign. I figure most people would use TS with a picture of a map, and the geospatial PDF would be only a small percent of users.

**Reply by Lee-Carre** [reply]
I think I have a better understanding, now. Thanks.

> If you already have an offline map downloaded, that is most likely the preferred solution — TS is a backup for when you only have access to a physical map or no internet.

I now recognise that I'm assuming that the route (of where to download data for) will be known in advance, and that there would be opportunity (time, network access, power, whatever) to actually download that data in regular map apps before departure.

Snapping a photo and setting a few calibration points will certainly be much quicker (and reliably so) than detailed technical prep.

I suppose I'm too used to surveying near civilisation (ample power supply, always some kind of network connectivity (nearby)), rather than being out in the wilderness (PV panels for charging, no network reception at all) 🙂. Perhaps also the Scout's motto of “be prepared” was drilled into me well, when young.

**Reply by kylecorry31** [reply]
Exactly! I also do a lot of prep before going on a new hike - but I know a lot of people who will show up at a new hiking location without a map of any kind haha

**Reply by Lee-Carre** [reply]
Follow-up:

> * estimating the dimensions (especially height) of large structures (e.g. buildings) (really, any distances beyond what's practical for a tape-measure, until I get a laser rangefinder)

> An even wilder idea (returning to surveying) would be to measure distance (in a single dimension) by computing linear displacement from the accelerometer data. A virtual, endless tape-measure of sorts, which should be more accurate over short distances than GNSS. Imagine lacking a laser rangefinder, and a sufficiently long tape-measure, while wanting to measure the length of a building-sized feature, or the width of a tunnel.

[StreetComplete v41](/streetcomplete/StreetComplete/releases/tag/v41.0) can measure distances via the camera (using Google's newfangled augmented-reality). I've yet to try it, though.

Thought it was a novel approach. I was reminded of Trail Sense using the camera as a sight for more accurate bearing measurement (for triangulation, which will be even more useful, now).

Though, a laser rangefinder is still on my techno-toys shopping list 😄.

**Comment by leavenode** [comment]
I don't typical have a SIM in my phones when I leave the house with them, so I like them to have as much offline capability as possible. Your app is brilliant for that - makes my phone into something like an electronic multitool.

**Reply by kylecorry31** [reply]
Thank you!

**Comment by Shadowstrike-code** [comment]
It's useful for everyday use

**Comment by MedwardR** [comment]
I am involved with Trail Life USA (similar to Boy Scouts) and I have found this app extremely useful during my various wilderness excursions. It's proven to be a reliable method of navigation in many scenarios, especially with its offline capabilities, and I could easily recommend it to anyone. Often I will plan out trips in Gaia and then import them into Trail Sense, which I use for all of my navigation during the trip. The photo maps feature is also very cool and I have used that as well.

This project is easily one of my favorites and I look forward to its further development!

**Reply by kylecorry31** [reply]
Thanks! I haven't used Gaia, is it easy to import the trips into Trail Sense (I'm assuming it's a GPX export from Gaia)?

**Reply by MedwardR** [reply]
Yes, Gaia allows you to export saved data (beacons, paths) as a GPX file, which can be easily imported into Trail Sense. This works with exporting a single path/beacon or many.

**Reply by kylecorry31** [reply]
Perfect, thank you - if there's any way you think I could make that process easier, let me know.

**Reply by MedwardR** [reply]
Actually, something Gaia lets you do is share GPX files into the app. It is already possible to share a raw location to Trail Sense, but then any other data (name, elevation, notes) must be entered manually.

Gaia can share paths/waypoints/folders as a GPX file, which would be a more seamless alternative to saving it in local storage and then opening Trail Sense and importing it.

If a file of an unknown type is shared to Gaia, it just throws an "unknown file type" pop-up.

https://github.com/kylecorry31/Trail-Sense/assets/117788102/a5921cbf-7d46-4328-9c49-9022face32eb

**Reply by kylecorry31** [reply]
Thank you for the suggestion, I've added 2 issues to track this functionality: https://github.com/kylecorry31/Trail-Sense/issues/2018 and https://github.com/kylecorry31/Trail-Sense/issues/2019

**Comment by Wumbo-11** [comment]
Hello developers, could you please tell me how the trail and beacon system works? I was quite confused when implementing the trail on my trips and hikes when it could not follow my GPS trail and was not displayed on the front of the compass circle.

I figured out how it works only through the action between two points when I tried to create a new trail. I mean, every time I stopped somewhere, I had to use a new marker (point) to be able to connect the previous trail with the current one _(and it was just a single line)._

How can I make it more flexible, aligned and updated according to the contours of the road and the GPS turned on, by creating a new line every few meters or every time? 
_(it is similar to a tracker system but without relying on the internet)._

I hope you understand what I'm saying, or you can make a button on the beacon and trail so that it can display the updated GPS trail every time. _(that would be very useful for exploring new places without losing the previous trail)._

**Thank you,
Greetings, Otcho.**

**Reply by kylecorry31** [reply]
Hi, I'm not sure I follow - I'll try to rephrase what I think you said: You want to record a trail and view it on the compass. The way you are currently doing that is by either using beacons or placing a waypoint on the path - both of which you are doing manually. You want this to be done automatically for you as you hike.

It sounds like you may be looking for the Backtrack feature. This feature will allow you to record a path in the background and you can adjust how often it does that. If you stop and start Backtrack, it will create a new path - so you won't lose the previous trail. I'm working on revamping all the guides this month, but here's a link to the current guide on Backtrack: https://github.com/kylecorry31/Trail-Sense/blob/fe533adace6ca20e710593b93556ad48988b9134/app/src/main/res/raw/navigate.md#how-to-trace-back-your-steps-backtrack 

If you want to see that Backtrack path while you hike, you can use the radar compass on the Navigation tab. This should be enabled by default, but if you had disabled it you can enable it on Settings > Navigation > Enable both "Show nearby beacons" and "Show nearby radar compass". On the same screen you can adjust the radius of what the radar compass shows. 

Let me know if that is what you were looking for - if not can you clarify further?

**Reply by Wumbo-11** [reply]
It worked perfectly, thank you for your instructions. I was very helpful for that.

**Reply by kylecorry31** [reply]
Glad that helped!

**Comment by vkuehn** [comment]
Hi,
Currently looking at some tracking apps which either are open source, work offline, could export tracks and hope fully could help in situations like hiken, boot travel etc. So came across yours
Looks promising so far.

**Reply by kylecorry31** [reply]
Thanks!
